src/manager.ts

```diff
@@ -1,550 +1,470 @@
+						this._previewManager.currentPanel?.close();
+					}
+
+					this._statusBar.removeServer(workspace?.uri);
+					this._serverGroupings.delete(workspace?.uri.toString());
+					if (this._serverGroupings.size === 0) {
+						this._statusBar.serverOff();
+						vscode.commands.executeCommand(
+							'setContext',
+							LIVE_PREVIEW_SERVER_ON,
+							false
+						);
+					}
+					this._connectionManager.removeConnection(workspace);
+				})
+			);
+			this._register(
+				serverGrouping.onShouldLaunchEmbeddedPreview((e) =>
+					this._previewManager.launchFileInEmbeddedPreview(
+						e.file,
+						e.relative,
+						e.panel,
+						e.connection
+					)
+				)
+			);
+			this._register(
+				serverGrouping.onShouldLaunchExternalPreview((e) =>
+					this._previewManager.launchFileInExternalBrowser(
+						e.file,
+						e.relative,
+						e.debug,
+						e.connection
+					)
+				)
+			);
+			this._serverGroupings.set(workspace?.uri.toString(), serverGrouping);
+		}
+
+		return serverGrouping;
+	}
+
+	private async _openPreview(
+		internal: boolean,
+		file: string,
+		serverGrouping: ServerGrouping,
+		isRelative: boolean,
+		debug = false
+	): Promise<void> {
+		if (internal) {
+			// for now, ignore debug or no debug for embedded preview
+			serverGrouping.createOrShowEmbeddedPreview(undefined, file, isRelative);
+		} else {
+			await serverGrouping.showPreviewInBrowser(file, isRelative, debug);
+		}
+	}
+
+	private _getFileInfo(
+		file: vscode.Uri | string | undefined,
+		fileStringRelative: boolean
+	): { filePath: string; isRelative: boolean } {
+		if (typeof file == 'string') {
+			return { filePath: file, isRelative: fileStringRelative };
+		} else if (file instanceof vscode.Uri) {
+			let filePath = file?.fsPath ?? file?.path;
+
+			if (!filePath) {
+				const activeFilePath =
+					vscode.window.activeTextEditor?.document.fileName;
+				if (activeFilePath) {
+					filePath = activeFilePath;
+					fileStringRelative = false;
+				}
+			}
+
+			return { filePath, isRelative: fileStringRelative };
+		} else {
+			const activeFilePath = vscode.window.activeTextEditor?.document.fileName;
+			if (activeFilePath) {
+				return { filePath: activeFilePath, isRelative: false };
+			}
+		}
+
+		return { filePath: '/', isRelative: fileStringRelative };
+	}
+
+	private _hasServerRunning(): boolean {
+		const isRunning = Array.from(this._serverGroupings.values()).filter(
+			(group) => group.running
+		);
+		return isRunning.length !== 0;
+	}
+
+	private _openNoTarget(): void {
+		// opens index at first open server or opens a loose workspace at root
+		const workspaces = vscode.workspace.workspaceFolders;
+		if (workspaces && workspaces.length > 0) {
+			for (let i = 0; i < workspaces.length; i++) {
+				const currWorkspace = workspaces[i];
+				const manager = this._serverGroupings.get(currWorkspace.uri.toString());
+				if (manager) {
+					vscode.commands.executeCommand(
+						`${SETTINGS_SECTION_ID}.start.preview.atFile`,
```

Comment:
I would avoid calling executeCommand on your own extension. I think it's cleaner to call a method normally (and you can also register a command that calls the same method.) With executeCommand you lose TS checks and it lets you call arbitrarily between two unrelated pieces of code.